### ISSUE TITLE
Manual requires-edge editing from the Graph view

### ISSUE THREAD

**Issue by LemonPhase** [LemonPhase/meno#14]
# Manual `requires` edge editing from the Graph view

## Problem Statement

Users can rename and delete Concepts, but the `requires` structure itself is
read-only: the only edges in the Graph are the ones investigation produced
(plus the remedial-splice edge from `insert_remedial`). When the model gets a
dependency wrong — says Attention requires Softmax when the user knows it
doesn't, or misses that Cross entropy rests on Softmax — the user has no way to
correct it. Since later Sessions' Paths are linearized from `requires`, a wrong
edge is not a cosmetic problem: it mis-orders everything the user learns next.
The Graph is meant to be user-curated (ADR-0003); right now the curation stops
at labels and membership.

## Solution

Add manual edge editing: from the Graph view (`/graph`), the user can add a
prerequisite edge between any two Concepts and remove any existing edge. Each
change is recorded append-only as an Edit (new kinds `link` / `unlink`) and fed
to later agent calls as context, exactly like renames and deletes. The one and
only bound is cycle rejection: an edge that would close a cycle in the
`requires` DAG is refused with a plain-language 409, because every consumer
(`linearizePath`, `computeDepths` in `dag-layout.ts`) assumes a DAG. Everything
else is allowed per ADR-0003's trust-the-user philosophy — no warning, no
confirmation, no other validation.

Vocabulary throughout follows CONTEXT.md: edge endpoints are the *dependent*
(the Concept whose `requires` gains/loses an entry) and the *prerequisite*.

## User Stories

1. As a learner, I want to mark one Concept as requiring another, so that I can
   correct a dependency the investigation missed.
2. As a learner, I want to remove a prerequisite edge that is wrong, so that my
   Graph says what I actually believe about how ideas depend on each other.
3. As a learner, I want a cycle-closing edge refused with a plain explanation,
   so that the graph stays a DAG I can trust to order my learning.
4. As a learner, I want adding an edge that already exists (or removing one
   that doesn't) to just succeed quietly, so that I'm never punished for
   clicking twice.
5. As a learner, I want my edge edits recorded in the same Edit log as renames
   and deletes and read by the agent, so that Meno respects how I organize my
   own understanding in future Graph updates.
6. As a learner, I want my in-progress Sessions left exactly as they are when I
   edit edges, so that curating the Graph never yanks a Path out from under me.

### Store (`src/lib/store.ts`)

- **New pure helper, next to `linearizePath`:**

  ```ts
  /**
   * Would adding `requiresId` to `dependentId`'s `requires` close a cycle?
   * Reachability over existing edges: true iff the prerequisite already
   * (transitively) requires the dependent.
   */
  export function wouldCreateCycle(
    concepts: Concept[],
    dependentId: string,
    requiresId: string,
  ): boolean;
  ```

  Depth-first walk from `requiresId` following `requires`; true if
  `dependentId` is reached. Ignore dangling ids (a requires entry pointing at
  a deleted Concept) the way `linearizePath` already does.

- **New mutators, mirroring `renameConcept`'s batch-and-append-only-Edit
  shape** (single Firestore batch: one concept update + one edits `set`):

  ```ts
  export async function linkConcepts(
    dependent: Concept,
    prerequisite: Concept,
    graphId?: string, // default DEMO_USER_ID, as everywhere
  ): Promise<void>;

  export async function unlinkConcepts(
    dependent: Concept,
    prerequisite: Concept,
    graphId?: string,
  ): Promise<void>;
  ```

  `linkConcepts` appends `prerequisite.id` to `dependent.requires`;
  `unlinkConcepts` filters it out. Callers guarantee both Concepts exist, the
  ids differ, and (for link) no cycle — the route enforces these.

- **Duplicate link / absent unlink are no-op successes, recorded as nothing.**
  If the edge already exists (or doesn't, on unlink), the function returns
  without writing — no concept update, no Edit. The log records changes, not
  intentions; a no-op Edit would also teach the agent nothing.

- **`Edit` type (`src/lib/types.ts`):** widen `kind` to
  `"rename" | "delete" | "link" | "unlink"`. For both new kinds:
  - `conceptId` = the dependent's id (the Concept whose `requires` changed),
  - `before` = the prerequisite's label,
  - `after` = the dependent's label (never null for link/unlink),
  - the kind carries the direction. Labels are point-in-time snapshots, same
    as rename/delete — later renames do not rewrite history.

- **`formatEditContext`:** add the two lines so the agent sees these
  curations too:
  - link: `` - linked "${before}" as a prerequisite of "${after}" ``
  - unlink: `` - unlinked "${before}" from "${after}" ``

### API (`src/app/api/concepts/[id]/requires/route.ts`, new file)

Consistent with the sibling `src/app/api/concepts/[id]/route.ts` (JSON body,
same error style, same response shape). `[id]` is the dependent.

- **`POST` — add an edge.** Body: `{ "requiresId": string }`.
  - 400 `{ error: "requiresId is required" }` when missing/not a string.
  - 400 when `requiresId === id` (self-edge), plain message, e.g.
    `"Attention" can't require itself`.
  - 404 `{ error: "no such Concept" }` when either endpoint is unknown (look
    both up via `getGraphOverview().concepts`, as `conceptById` does today).
  - 409 when `wouldCreateCycle` is true, plain language naming both Concepts,
    e.g. `"Dot product" already depends on "Attention", so it can't also be
    its prerequisite`. Nothing is written — no edge, no Edit.
  - 200 on success (including the duplicate-edge no-op); body is
    `getSessionState(sessionIdFrom(request, body))`, exactly like the
    existing PATCH — the Graph page ignores the body and refetches
    `/api/graph`, same as it does for rename/delete today.
- **`DELETE` — remove an edge.** Body: `{ "requiresId": string }`. Same 400s
  and 404s; no 409 case (removing an edge cannot create a cycle). 200 on
  success (including the absent-edge no-op); body is
  `getSessionState(sessionIdFrom(request))`, like the existing DELETE.

### No retroactive Path changes

Existing Sessions' Paths were linearized when their diagnosis was applied and
stay exactly as they are. A new edge affects only *future* investigations'
linearization; removing an edge does not reorder, prune, or unblock anything
in progress. Edging a Concept that sits on an in-progress Path — even the
Active one — is allowed and mutates nothing but that Concept's `requires`.
(Only deletion has the Active-Concept bound, and this is not deletion.)

### UI (`src/components/GraphView.tsx`, `src/app/graph/page.tsx`)

The detail panel already handles selection, rename, and delete; edge editing
extends it rather than inventing a new surface.

- **New optional props**, same shape as `onRename`/`onDelete`:

  ```ts
  onLink?: (conceptId: string, requiresId: string) => Promise<void> | void;
  onUnlink?: (conceptId: string, requiresId: string) => Promise<void> | void;
  ```

  (The session minimap and any other GraphView consumer simply don't pass
  them, and get today's read-only panel.)

- **"Requires" block in the detail panel** (rendered when `onLink`/`onUnlink`
  are provided), between the status line and the `.acts` buttons:
  - One row per current prerequisite: its label plus a small ✕ that calls
    `onUnlink(selected.id, requiresId)`. No confirmation — unlinking is
    reversible and ADR-0003 trusts the user.
    - Resolve labels via `concepts.find((c) => c.id === requiresId)`; skip
      dangling ids (deleted prerequisites leave no entry anyway after
      `deleteConcept`'s cleanup, but be tolerant).
    - When the Concept requires nothing: one quiet line, "Requires nothing."
  - An "Add prerequisite" button. Clicking it enters *linking mode*: the
    panel shows a hint — `Click the concept that "Attention" requires` — and
    the next node click on the canvas calls `onLink(selected.id, clickedId)`
    instead of opening that node's panel, then exits linking mode. Clicking
    the selected node itself or pressing Escape cancels (hook into the
    existing Escape handler in `GraphView`). While linking, give the canvas a
    class (e.g. `gcanvas linking`) so nodes can show a pointer affordance;
    keep styling minimal and in the file's existing idiom.
- **Errors surface through the page's existing toast.** A rejected (409)
  link throws out of the page's `edit()` helper and lands in the same
  dismissable error toast rename/delete already use. No new error UI.
- **Graph page wiring** (`src/app/graph/page.tsx`):

  ```ts
  onLink={(id, requiresId) =>
    edit(`/api/concepts/${id}/requires`, "POST", { requiresId })}
  onUnlink={(id, requiresId) =>
    edit(`/api/concepts/${id}/requires`, "DELETE", { requiresId })}
  ```

  The existing `edit()` helper already supports a body on any method and
  calls `refresh()` + `announceSessionsChanged()` on success — no changes to
  it.
- **"Your edits" log on the graph page** currently renders only rename and
  delete — extend it for the two new kinds, e.g.
  `<b>{e.before}</b> now required by <b>{e.after}</b>` and
  `<b>{e.before}</b> no longer required by <b>{e.after}</b>`, and update the
  empty-state copy ("Renaming, removing, or re-linking a concept is recorded
  here…").

### Edge cases (all decided, none deferred)

- **Self-edge:** 400, never written, never recorded.
- **Duplicate link / absent unlink:** 200 no-op, no Edit.
- **Cycle:** 409 with both labels named; the Graph stays untouched.
- **Edging an Unlocked Concept, or one on an in-progress Path (Active
  included):** allowed; no status, Path, or Lesson changes.
- **Removing a dependent's last edge:** fine — it becomes a root (depth 0)
  and the layout re-layers it on the next render, no special casing.
- **Linking two Concepts from different Sessions:** fine — `requires` is a
  Graph fact, not a Session fact (ADR-0002/0004).
- **Either endpoint is remedial or Skipped:** irrelevant; edges don't care.

### Doc updates

- **CONTEXT.md — Requires entry:** append manual editing and its one bound.
  Suggested wording: "The user can also add and remove edges by hand from the
  Graph view; the only refusal is an edge that would close a cycle, because
  linearization and layout assume a DAG."
- **CONTEXT.md — Edit entry:** change "(rename or delete)" to "(rename,
  delete, link, unlink)" and note that link/unlink record the two endpoint
  labels (prerequisite in `before`, dependent in `after`).
- **README — Status section:** drop "manual edges" from the future-work list
  ("richer graph editing (merge nodes)" remains) and extend the
  review-the-graph feature bullet: "You can rename, delete, or re-link nodes
  yourself…"

## Testing Decisions

New black-box tests in `tests/requires.test.ts`, following the pattern of
`tests/edits.test.ts`: drive the scripted fake through
`startInvestigatedSession()` / `reachLearning()` (whose EXTRACTION already
builds the chain Dot product ← Softmax ← Attention), call the route handlers
directly, assert on the response and on stored state (`graphRef()` reads,
`storedEdits()`-style helper). No test reaches into `wouldCreateCycle` or
other internals.

- **Link adds the edge and records the Edit:** note that in the single-Session
  seed (Dot product ← Softmax ← Attention, with Attention also requiring
  Dot product) every missing edge is a cycle, so use
  `startOverlappingSession()` first — it adds Cross entropy (requires Softmax
  only). Then POST dependent=Cross entropy, requiresId=Dot product is a fresh,
  acyclic edge. Assert 200, the stored Concept's `requires` contains the new
  id, and one Edit with `kind: "link"`, `conceptId` = dependent id,
  `before` = prerequisite label, `after` = dependent label.
- **Cycle rejected:** POST dependent=Dot product, requiresId=Attention → 409,
  error names both labels; stored `requires` unchanged; zero Edits.
- **Self-edge:** POST with `requiresId === id` → 400; nothing written, no
  Edit.
- **Unknown Concepts:** POST/DELETE with an unknown dependent → 404; unknown
  `requiresId` → 404; missing `requiresId` → 400. No Edits recorded.
- **Duplicate link is a quiet no-op:** POST Attention requires Softmax
  (already an edge) → 200; `requires` still contains the id exactly once;
  zero Edits.
- **Unlink removes the edge and records the Edit:** DELETE Attention requires
  Softmax → 200; edge gone; Edit with `kind: "unlink"`, `before: "Softmax"`,
  `after: "Attention"`.
- **Absent unlink is a quiet no-op:** DELETE Dot product requires Softmax →
  200; zero Edits.
- **No Path mutation:** after `reachLearning()`, link or unlink any edge and
  assert the returned `state.session.path` and each Concept's `status` are
  unchanged.
- **Edits reach later agent calls:** mirror the existing context tests —
  after a link (Cross entropy → Dot product, as above), the next grading
  prompt contains
  `linked "Dot product" as a prerequisite of "Cross entropy"`;
  after an unlink, the next investigation prompt contains
  `unlinked "Softmax" from "Attention"`.

No UI/component tests, per the repo's testing philosophy — the panel
interaction is verified by hand with `npm run seed` in the emulator.

## Out of Scope

- **Merging Concepts** — the sibling "merge nodes" feature; its own spec.
- Edge editing from the Session view or minimap — Graph view only.
- Agent-suggested edges, or the agent reading link/unlink Edits as anything
  beyond the existing plain context lines.
- Any retroactive Path reordering, re-linearization of in-progress Sessions,
  or notifications that "this Path is now stale."
- Confirmation dialogs, undo, and any validation beyond self-edge and cycle
  rejection (ADR-0003).
- Batch edge editing, drag-to-connect canvas interactions.

## Further Notes

- `linearizePath`'s cycle-breaking fallback comment ("a model could produce
  one") stays honest — the guard remains for model-produced cycles; manual
  edges simply can no longer be a source of them. `computeDepths`'s visiting
  guard likewise stays.
- The delete-time cleanup (removing the deleted Concept from dependents'
  `requires`) already keeps `requires` dangling-free; edge editing neither
  adds nor removes that obligation.
- Watch the seeded graph after implementation: `npm run seed` should gain a
  link/unlink Edit or two only if that helps exercise the UI states — not
  required.
- Verify before claiming done: `npm run lint`, `npx tsc --noEmit`, `npm test`.
